File: backend/src/repositories/progress.repository.ts
import { BaseRepository } from './base.repository';
import { db } from '../config/firebase';

export interface UserProgress {
  id: string;
  user_id: string;
  course_id: string;
  module_id: string;
  completed: boolean;
  completed_at?: string;
  progress_percentage: number; // 0-100
  last_accessed_at: string;
  time_spent?: number;
  video_time_watched?: number;
  video_duration?: number;
  times_accessed?: number;
}

export interface CourseProgress {
  id: string;
  user_id: string;
  course_id: string;
  total_modules: number;
  completed_modules: number;
  progress_percentage: number; // 0-100
  last_accessed_at: string;
  updated_at: string;
}

export class UserProgressRepository extends BaseRepository<UserProgress> {
  constructor() {
    super('user_progress');
  }

  /**
   * Buscar progreso de un usuario en un módulo específico
   */
  async findByUserAndModule(userId: string, courseId: string, moduleId: string): Promise<UserProgress | null> {
    try {
      const snapshot = await this.collection
        .where('user_id', '==', userId)
        .where('course_id', '==', courseId)
        .where('module_id', '==', moduleId)
        .limit(1)
        .get();
      
      if (snapshot.empty) return null;
      
      const doc = snapshot.docs[0];
      return { id: doc.id, ...doc.data() } as UserProgress;
    } catch (error) {
      console.error('Error getting user progress:', error);
      throw error;
    }
  }

  /**
   * Obtener todo el progreso de un usuario en un curso
   */
  async findByUserAndCourse(userId: string, courseId: string): Promise<UserProgress[]> {
    try {
      const snapshot = await this.collection
        .where('user_id', '==', userId)
        .where('course_id', '==', courseId)
        .get();
      
      return snapshot.docs.map(doc => ({
        id: doc.id,
        ...doc.data()
      } as UserProgress));
    } catch (error) {
      console.error('Error getting user progress for course:', error);
      throw error;
    }
  }

  /**
   * Guardar o actualizar progreso de acceso a módulo
   */
  async saveAccess(
    userId: string,
    courseId: string,
    moduleId: string,
    progressPercentage?: number
  ): Promise<void> {
    try {
      const existing = await this.findByUserAndModule(userId, courseId, moduleId);
      const now = new Date().toISOString();

      if (existing) {
        await this.collection.doc(existing.id).update({
          last_accessed_at: now,
          times_accessed: (existing.times_accessed || 0) + 1,
          ...(progressPercentage !== undefined && {
            progress_percentage: Math.max(existing.progress_percentage || 0, progressPercentage)
          })
        });
      } else {
        await this.collection.add({
          user_id: userId,
          course_id: courseId,
          module_id: moduleId,
          completed: false,
          progress_percentage: progressPercentage || 0,
          last_accessed_at: now,
          times_accessed: 1
        });
      }
    } catch (error) {
      console.error('Error saving module access:', error);
      throw error;
    }
  }

  /**
   * Guardar progreso parcial (videos, tiempo, etc.)
   */
  async saveProgress(
    userId: string,
    courseId: string,
    moduleId: string,
    progressData: {
      progress_percentage?: number;
      video_time_watched?: number;
      video_duration?: number;
      time_spent?: number;
    }
  ): Promise<void> {
    try {
      const existing = await this.findByUserAndModule(userId, courseId, moduleId);
      const now = new Date().toISOString();

      if (existing) {
        const updates: any = {
          last_accessed_at: now,
          ...progressData
        };

        // Si se proporciona tiempo de video, calcular porcentaje
        if (progressData.video_time_watched && progressData.video_duration) {
          const calculatedPercentage = Math.min(
            100,
            Math.round((progressData.video_time_watched / progressData.video_duration) * 100)
          );
          updates.progress_percentage = Math.max(
            existing.progress_percentage || 0,
            calculatedPercentage
          );
        }

        await this.collection.doc(existing.id).update(updates);
      } else {
        const progressPercentage = progressData.progress_percentage ||
          (progressData.video_time_watched && progressData.video_duration
            ? Math.round((progressData.video_time_watched / progressData.video_duration) * 100)
            : 0);

        await this.collection.add({
          user_id: userId,
          course_id: courseId,
          module_id: moduleId,
          completed: false,
          progress_percentage: progressPercentage,
          last_accessed_at: now,
          times_accessed: 1,
          ...progressData
        });
      }
    } catch (error) {
      console.error('Error saving module progress:', error);
      throw error;
    }
  }

  /**
   * Marcar módulo como completado
   */
  async markComplete(userId: string, courseId: string, moduleId: string): Promise<void> {
    try {
      const existing = await this.findByUserAndModule(userId, courseId, moduleId);
      const now = new Date().toISOString();

      if (existing) {
        await this.collection.doc(existing.id).update({
          completed: true,
          completed_at: now,
          progress_percentage: 100,
          last_accessed_at: now
        });
      } else {
        await this.collection.add({
          user_id: userId,
          course_id: courseId,
          module_id: moduleId,
          completed: true,
          completed_at: now,
          progress_percentage: 100,
          last_accessed_at: now,
          times_accessed: 1
        });
      }
    } catch (error) {
      console.error('Error marking module complete:', error);
      throw error;
    }
  }
}

export class CourseProgressRepository extends BaseRepository<CourseProgress> {
  constructor() {
    super('course_progress');
  }

  /**
   * Obtener progreso de curso de un usuario
   */
  async findByUserAndCourse(userId: string, courseId: string): Promise<CourseProgress | null> {
    try {
      const snapshot = await this.collection
        .where('user_id', '==', userId)
        .where('course_id', '==', courseId)
        .limit(1)
        .get();

      if (snapshot.empty) return null;

      const doc = snapshot.docs[0];
      return { id: doc.id, ...doc.data() } as CourseProgress;
    } catch (error) {
      console.error('Error getting course progress:', error);
      throw error;
    }
  }

  /**
   * Actualizar progreso del curso
   */
  async updateProgress(
    userId: string,
    courseId: string,
    totalModules: number,
    completedModules: number
  ): Promise<void> {
    try {
      const existing = await this.findByUserAndCourse(userId, courseId);
      const now = new Date().toISOString();
      const progressPercentage = totalModules > 0
        ? Math.round((completedModules / totalModules) * 100)
        : 0;

      if (existing) {
        await this.collection.doc(existing.id).update({
          total_modules: totalModules,
          completed_modules: completedModules,
          progress_percentage: progressPercentage,
          last_accessed_at: now,
          updated_at: now
        });
      } else {
        await this.collection.add({
          user_id: userId,
          course_id: courseId,
          total_modules: totalModules,
          completed_modules: completedModules,
          progress_percentage: progressPercentage,
          last_accessed_at: now,
          updated_at: now
        });
      }
    } catch (error) {
      console.error('Error updating course progress:', error);
      throw error;
    }
  }
}

File: backend/app/repositories/progress_repository.py
"""Progress repository for Firestore access."""

from app.firebase import get_db


class ProgressRepository:
    """Data access for user_progress and course_progress collections."""

    def __init__(self) -> None:
        self._db = get_db()

    # --- User progress ---

    def get_module_progress(self, user_id: str, course_id: str, module_id: str) -> dict | None:
        query = (
            self._db.collection("user_progress")
            .where("user_id", "==", user_id)
            .where("course_id", "==", course_id)
            .where("module_id", "==", module_id)
            .limit(1)
            .stream()
        )
        doc = next(iter(query), None)
        if doc is None:
            return None
        return self._doc_to_dict(doc)

    def list_module_progress(self, user_id: str, course_id: str) -> list[dict]:
        query = (
            self._db.collection("user_progress")
            .where("user_id", "==", user_id)
            .where("course_id", "==", course_id)
            .stream()
        )
        return [self._doc_to_dict(doc) for doc in query]

    def save_module_progress(self, user_id: str, course_id: str, module_id: str, payload: dict) -> None:
        existing = self.get_module_progress(user_id, course_id, module_id)
        collection = self._db.collection("user_progress")
        if existing:
            doc_ref = collection.document(existing["id"])
            doc_ref.update(payload)
        else:
            doc_ref = collection.document()
            payload = {
                **payload,
                "user_id": user_id,
                "course_id": course_id,
                "module_id": module_id,
            }
            doc_ref.set(payload)

    # --- Course progress ---

    def get_course_progress(self, user_id: str, course_id: str) -> dict | None:
        query = (
            self._db.collection("course_progress")
            .where("user_id", "==", user_id)
            .where("course_id", "==", course_id)
            .limit(1)
            .stream()
        )
        doc = next(iter(query), None)
        if doc is None:
            return None
        return self._doc_to_dict(doc)

    def save_course_progress(self, user_id: str, course_id: str, payload: dict) -> None:
        existing = self.get_course_progress(user_id, course_id)
        collection = self._db.collection("course_progress")
        if existing:
            collection.document(existing["id"]).update(payload)
        else:
            doc_ref = collection.document()
            payload = {
                **payload,
                "user_id": user_id,
                "course_id": course_id,
            }
            doc_ref.set(payload)

    @staticmethod
    def _doc_to_dict(doc) -> dict:
        data = doc.to_dict()
        data["id"] = doc.id
        return data
